agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java

```diff
@@ -0,0 +1,261 @@
+    addThrowable(span, unwrapThrowable(throwable));
+  }
+
+  // TODO semantic attributes
+  public static void addThrowable(final Span span, final Throwable throwable) {
+    span.setAttribute(MoreTags.ERROR_MSG, throwable.getMessage());
+    span.setAttribute(MoreTags.ERROR_TYPE, throwable.getClass().getName());
+
+    final StringWriter errorString = new StringWriter();
+    throwable.printStackTrace(new PrintWriter(errorString));
+    span.setAttribute(MoreTags.ERROR_STACK, errorString.toString());
+  }
+
+  public Span getCurrentSpan() {
+    return tracer.getCurrentSpan();
+  }
+
+  public Scope newScope(Span span) {
+    return tracer.withSpan(span);
+  }
+
+  // TODO should end methods remove SPAN attribute from request as well?
+  public void end(Span span, Scope scope, int responseCode) {
+    if (scope == null) {
+      return;
+    }
+
+    if (span != null) {
+      end(span, responseCode);
+    }
+
+    scope.close();
+  }
+
+  public void end(Span span, int responseStatus) {
+    setStatus(span, responseStatus);
+    span.end();
+  }
+
+  public void endExceptionally(Span span, Scope scope, Throwable throwable, int responseStatus) {
+    if (scope == null) {
+      return;
+    }
+
+    if (span != null) {
+      endExceptionally(span, throwable, responseStatus);
+    }
+
+    scope.close();
+  }
+
+  public void endExceptionally(Span span, Throwable throwable, int responseStatus) {
+    if (responseStatus == 200) {
+      // TODO I think this is wrong.
+      // We must report that response status that was actually sent to end user
+      // We may change span status, but not http_status attribute
+      responseStatus = 500;
+    }
+    setStatus(span, responseStatus);
+    onError(span, unwrapThrowable(throwable));
+    span.end();
+  }
+
+  protected Throwable unwrapThrowable(Throwable throwable) {
+    return throwable instanceof ExecutionException ? throwable.getCause() : throwable;
+  }
+
+  private <C> SpanContext extract(final C carrier, final HttpTextFormat.Getter<C> getter) {
+    final Context context =
+        getPropagators().getHttpTextFormat().extract(Context.current(), carrier, getter);
+    final Span span = getSpan(context);
+    return span.getContext();
+  }
+
+  private void setStatus(Span span, int status) {
+    SemanticAttributes.HTTP_STATUS_CODE.set(span, status);
+    // TODO status_message
+    if (Config.get().getHttpServerErrorStatuses().get(status)) {
+      span.setStatus(Status.UNKNOWN);
+    }
+  }
+
+  protected abstract Integer peerPort(REQUEST request);
+
+  protected abstract String peerHostIP(REQUEST request);
+
+  protected abstract HttpTextFormat.Getter<REQUEST> getGetter();
+
+  protected abstract URI url(REQUEST request) throws URISyntaxException;
+
+  protected abstract String method(REQUEST request);
+
+  protected abstract void attachSpanToRequest(Span span, REQUEST request);
+
+  protected abstract Span getAttachedSpan(REQUEST request);
```

Comment:
same comment as above
```suggestion
  protected Span getAttachedSpan(REQUEST request) {
    return null;
  }
```